### DISCUSSION TITLE
Failed to remove "point label" actor properly

### DISCUSSION THREAD

**Discussion by machur** [pyvista/pyvista#3829] [Q&A]
Hi, I'm working on changing the visibility of point labels. I've tried to hide them by removing actor acquired from a call to add_point_labels() of the plotter, but not all elements have been removed. 

The visualizations of this issue have been generated based on the example code for point label usage from https://docs.pyvista.org/examples/02-plot/labels.html:

![image](https://user-images.githubusercontent.com/4650515/212566033-e4024191-14d9-40a0-b7f2-b9e77349f44a.png)

Once the actor is removed, some remainings are still being rendered (white little boxes in the corners):

![image](https://user-images.githubusercontent.com/4650515/212566015-e904de4a-4fb1-475f-bc6b-153a893cbc47.png)

Does anyone have any idea how to tackle that issue and remove the labels entirely?

Code snippet to reproduce the issue:

```
import pyvista as pv
from pyvista import examples

grid = pv.UnstructuredGrid(examples.hexbeamfile)

plotter = pv.Plotter()
plotter.add_mesh(grid, show_edges=True, color="tan")

points = grid.points
mask = points[:, 0] == 0
actor = plotter.add_point_labels(points[mask], points[mask].tolist(), point_size=20, font_size=36)
plotter.camera_position = [(-1.5, 1.5, 3.0), (0.05, 0.6, 1.2), (0.2, 0.9, -0.25)]

plotter.remove_actor(actor)

plotter.show()
```

**Comment by MatthewFlamm** [comment]
The default for `show_points` argument is `True`, which means that there are actually two separate actors that are plotted here, yet only one is returned.  This is awkward API for this use case, but it is useful for quickly plotting labels themselves with the points together.  I recommend setting `show_points=False` and then manually plotting the points for the labels.  You can store the actors for each and remove individually.

I wonder if this should be mentioned somehow in the documentation.

**Comment by banesullivan** [comment]
IMO, since `add_point_labels()` adds two actors when `show_points=True`, it should return those two actors:

https://github.com/pyvista/pyvista/blob/728f493b4968b86576ed2b964415baf5eba6b07d/pyvista/plotting/plotting.py#L4915

This would be an easy fix, though while we're doing this, we should make sure any other methods that add multiple actors return all of the actors

**Reply by banesullivan** [reply]
This diff would do it:

```diff
diff --git a/pyvista/plotting/plotting.py b/pyvista/plotting/plotting.py
index a6d38004..8ba33662 100644
--- a/pyvista/plotting/plotting.py
+++ b/pyvista/plotting/plotting.py
@@ -4790,6 +4790,9 @@ class BasePlotter(PickingHelper, WidgetHelper):
         vtk.vtkActor2D
             VTK label actor.  Can be used to change properties of the labels.

+        pyvista.Actor
+            The points actor if ``show_points`` is ``True``.
+
         Examples
         --------
         >>> import numpy as np
@@ -4896,9 +4899,13 @@ class BasePlotter(PickingHelper, WidgetHelper):
         self.remove_actor(f'{name}-points', reset_camera=False)
         self.remove_actor(f'{name}-labels', reset_camera=False)

+        label_actor = _vtk.vtkActor2D()
+        label_actor.SetMapper(labelMapper)
+        self.add_actor(label_actor, reset_camera=False, name=f'{name}-labels', pickable=False)
+
         # add points
         if show_points:
-            self.add_mesh(
+            points_actor = self.add_mesh(
                 algo or points,
                 color=point_color,
                 point_size=point_size,
@@ -4908,10 +4915,8 @@ class BasePlotter(PickingHelper, WidgetHelper):
                 reset_camera=reset_camera,
                 render=render,
             )
+            return label_actor, points_actor

-        label_actor = _vtk.vtkActor2D()
-        label_actor.SetMapper(labelMapper)
-        self.add_actor(label_actor, reset_camera=False, name=f'{name}-labels', pickable=False)
         return label_actor

     def add_point_scalar_labels(self, points, labels, fmt=None, preamble='', **kwargs):
```

**Reply by MatthewFlamm** [reply]
Given the API as it is, this is probably best.

**Reply by machur** [reply]
Thanks for the quick responses. @MatthewFlamm setting the flag show_points to false was good enough for my use-case. I wish the approach suggested by @banesullivan was incorporated into pyvista though. It fits nicely into the existing actor-based API. Feel free to turn this discussion into an opened issue/PR.
